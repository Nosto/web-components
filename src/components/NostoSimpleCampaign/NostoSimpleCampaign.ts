import { customElement } from "../decorators"
import { nostojs } from "@nosto/nosto-js"
import { JSONResult, JSONProduct } from "@nosto/nosto-js/client"
import { NostoElement } from "../NostoElement"
import { addRequest } from "../NostoCampaign/orchestrator"

/**
 * A Shopify-specific custom element that renders a Nosto campaign with different display modes.
 * This component fetches campaign data from Nosto and renders it in grid, carousel, or bundle layout.
 * It can optionally delegate product card rendering to Shopify via NostoDynamicCard.
 *
 * @property {string} placement - The placement identifier for the campaign. Required.
 * @property {string} [mode] - The rendering mode: 'grid', 'carousel', or 'bundle'. Defaults to 'grid'.
 * @property {string} [card] - If provided, delegates card rendering to Shopify via NostoDynamicCard.
 *
 * @example
 * ```html
 * <!-- Basic grid layout -->
 * <nosto-simple-campaign placement="frontpage-nosto-1"></nosto-simple-campaign>
 *
 * <!-- Carousel layout with Shopify card rendering -->
 * <nosto-simple-campaign
 *   placement="frontpage-nosto-2"
 *   mode="carousel"
 *   card="product-card">
 * </nosto-simple-campaign>
 * ```
 */
@customElement("nosto-simple-campaign")
export class NostoSimpleCampaign extends NostoElement {
  /** @private */
  static attributes = {
    placement: String,
    mode: String,
    card: String,
    productId: String,
    variantId: String
  }

  placement!: string
  mode?: string
  card?: string
  productId?: string
  variantId?: string

  async connectedCallback() {
    if (!this.placement) {
      throw new Error("placement attribute is required for NostoSimpleCampaign")
    }

    this.toggleAttribute("loading", true)
    try {
      await this.#initializeCampaign()
    } finally {
      this.toggleAttribute("loading", false)
    }
  }

  async #initializeCampaign() {
    const api = await new Promise(nostojs)
    const rec = (await addRequest({
      placement: this.placement,
      productId: this.productId,
      variantId: this.variantId,
      responseMode: "JSON_ORIGINAL"
    })) as JSONResult

    if (!rec) {
      return
    }

    const mode = this.mode || "grid"
    const markup = await this.#renderCampaign(rec, mode)
    this.innerHTML = markup
    api.attributeProductClicksInCampaign(this, rec)
  }

  async #renderCampaign(rec: JSONResult, mode: string): Promise<string> {
    const products = (rec.products || []) as JSONProduct[]

    if (products.length === 0) {
      return ""
    }

    if (this.card) {
      return this.#renderWithDynamicCards(products, mode)
    } else {
      return this.#renderWithBasicCards(products, mode)
    }
  }

  #renderWithDynamicCards(products: JSONProduct[], mode: string): string {
    const containerClass = this.#getContainerClass(mode)
    const productCards = products
      .map((product: JSONProduct) => {
        const handle = this.#extractProductHandle(product)
        if (!handle) return ""

        return `<nosto-dynamic-card
            handle="${handle}"
            template="${this.card}"
            ${this.variantId ? `variant-id="${this.variantId}"` : ""}
          ></nosto-dynamic-card>`
      })
      .filter(card => card !== "")
      .join("")

    return `<div class="nosto-simple-campaign nosto-${mode}">
        ${containerClass ? `<div class="${containerClass}">` : ""}
          ${productCards}
        ${containerClass ? "</div>" : ""}
      </div>`
  }

  #renderWithBasicCards(products: JSONProduct[], mode: string): string {
    const containerClass = this.#getContainerClass(mode)
    const productCards = products
      .map((product: JSONProduct) => {
        return `<div class="nosto-product-card">
        <div class="nosto-product-image">
          ${product.thumb_url ? `<img src="${product.thumb_url}" alt="${product.name || ""}" />` : ""}
        </div>
        <div class="nosto-product-info">
          <h3 class="nosto-product-title">${product.name || ""}</h3>
          <div class="nosto-product-price">${product.price_text || ""}</div>
        </div>
      </div>`
      })
      .join("")

    return `<div class="nosto-simple-campaign nosto-${mode}">${containerClass ? `<div class="${containerClass}">` : ""}${productCards}${containerClass ? "</div>" : ""}</div>`
  }

  #getContainerClass(mode: string): string {
    switch (mode) {
      case "carousel":
        return "nosto-carousel-container"
      case "bundle":
        return "nosto-bundle-container"
      case "grid":
      default:
        return "nosto-grid-container"
    }
  }

  #extractProductHandle(product: JSONProduct): string | null {
    // Try to extract Shopify product handle from various possible fields
    if (product.url) {
      // Extract handle from URL like "/products/product-handle", ignoring query params and hash
      const match = product.url.match(/\/products\/([^/?#]+)/)
      if (match) return match[1]
    }

    // Fallback to handle or product_id if available
    return product.handle || product.product_id || null
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "nosto-simple-campaign": NostoSimpleCampaign
  }
}
