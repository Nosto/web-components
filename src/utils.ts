export function intersectionOf(...arrays: string[][]) {
  if (arrays.length === 0) {
    return []
  }
  return arrays?.reduce((intersection, currentArray) => {
    return currentArray.filter(item => intersection.includes(item))
  })
}

export function assertRequired<T>(object: T, ...properties: (keyof T & string)[]) {
  properties.forEach(property => {
    if (object[property] === undefined || object[property] === null) {
      throw new Error(`Property ${property} is required.`)
    }
  })
}

export function toCamelCase(str: string) {
  return str.replace(/-([a-z])/g, (_, l) => l.toUpperCase())
}

// Section Rendering API utilities
import { JSONResult } from "@nosto/nosto-js/client"

/**
 * Interface for elements that have a section property
 */
export interface SectionElement {
  section: string
}

/**
 * Fetches section markup using Shopify's Section Rendering API
 * @param sectionId - The section ID to render
 * @param searchQuery - The search query (typically product handles joined with ":")
 * @returns Promise that resolves to the rendered HTML markup
 * @throws Error if the fetch fails
 */
export async function fetchSectionMarkup(sectionId: string, searchQuery: string): Promise<string> {
  const target = new URL("/search", window.location.href)
  target.searchParams.set("section_id", sectionId)
  target.searchParams.set("q", searchQuery)

  const result = await fetch(target)
  if (!result.ok) {
    throw new Error(`Failed to fetch section ${sectionId}`)
  }

  return await result.text()
}

/**
 * Fetches product markup using Shopify's Section Rendering API
 * @param handle - The product handle
 * @param sectionId - The section ID to render
 * @param variantId - Optional variant ID
 * @returns Promise that resolves to the rendered HTML markup
 * @throws Error if the fetch fails
 */
export async function fetchProductSectionMarkup(
  handle: string,
  sectionId: string,
  variantId?: string
): Promise<string> {
  const params = new URLSearchParams()
  params.set("section_id", sectionId)
  if (variantId) {
    params.set("variant", variantId)
  }

  const result = await fetch(`/products/${handle}?${params}`)
  if (!result.ok) {
    throw new Error("Failed to fetch product data")
  }

  return await result.text()
}

/**
 * Extracts HTML content from the first element in the body of parsed HTML
 * @param html - The HTML string to parse
 * @returns The inner HTML of the first body element, or the original HTML if parsing fails
 */
export function extractSectionContent(html: string): string {
  const parser = new DOMParser()
  const doc = parser.parseFromString(html, "text/html")
  return doc.body.firstElementChild?.innerHTML?.trim() || html
}

/**
 * Replaces title content in elements with nosto-title attribute
 * @param html - The HTML string to process
 * @param title - The title to set
 * @returns The processed HTML string
 */
export function replaceTitleInMarkup(html: string, title: string): string {
  const parser = new DOMParser()
  const doc = parser.parseFromString(html, "text/html")

  const headingEl = doc.querySelector("[nosto-title]")
  if (headingEl) {
    headingEl.textContent = title
  }

  return doc.body.firstElementChild?.innerHTML?.trim() || html
}

/**
 * Gets section markup for a campaign using the Section Rendering API
 * @param element - Element with section property
 * @param rec - JSON result containing products and optional title
 * @returns Promise that resolves to the processed HTML markup
 */
export async function getCampaignSectionMarkup(element: SectionElement, rec: JSONResult): Promise<string> {
  const handles = rec.products.map(product => product.handle).join(":")
  const sectionHtml = await fetchSectionMarkup(element.section, handles)
  let markup = extractSectionContent(sectionHtml)

  if (rec.title) {
    markup = replaceTitleInMarkup(`<body><div>${markup}</div></body>`, rec.title)
  }

  return markup
}
