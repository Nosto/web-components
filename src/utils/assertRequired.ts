export function assertRequired<T>(object: T, ...properties: (keyof T & string)[]) {
  properties.forEach(property => {
    if (object[property] === undefined || object[property] === null) {
      throw new Error(`Property ${property} is required.`)
    }
  })
}
