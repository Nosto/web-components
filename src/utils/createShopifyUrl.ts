export function createShopifyUrl(url: string) {
  const root = new URL(window.Shopify?.routes?.root ?? "/", window.location.href).href
    if (root.endsWith("/") && url.startsWith("/")) {
    return new URL(root + url.slice(1))
  }
  return new URL(root + url)
}
