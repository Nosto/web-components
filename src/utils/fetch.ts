interface FetchOptions {
  cached?: boolean
}

/**
 * Internal function to handle common fetch logic with error checking.
 * @param url - The URL to fetch
 * @returns Promise that resolves to the Response object
 * @throws Error if the fetch request fails
 */
async function fetchWithErrorHandling(url: string) {
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(`Failed to fetch ${url}: ${response.status} ${response.statusText}`)
  }
  return response
}

// In-memory caches for getText and getJSON
const textCache = new Map<string, string>()
const jsonCache = new Map<string, unknown>()

/**
 * Fetches a URL and returns the response as text.
 * @param url - The URL to fetch
 * @param options - Optional configuration object
 * @param options.cached - If true, uses in-memory cache for the response
 * @returns Promise that resolves to the response text
 * @throws Error if the fetch request fails
 */
export async function getText(url: string, options?: FetchOptions): Promise<string> {
  if (options?.cached && textCache.has(url)) {
    return textCache.get(url)!
  }

  const response = await fetchWithErrorHandling(url)
  const text = await response.text()

  if (options?.cached) {
    textCache.set(url, text)
  }

  return text
}

/**
 * Fetches a URL and returns the response as a JSON object.
 * @param url - The URL to fetch
 * @param options - Optional configuration object
 * @param options.cached - If true, uses in-memory cache for the response
 * @returns Promise that resolves to the parsed JSON response
 * @throws Error if the fetch request fails or JSON parsing fails
 */
export async function getJSON<T = unknown>(url: string, options?: FetchOptions): Promise<T> {
  if (options?.cached && jsonCache.has(url)) {
    return jsonCache.get(url) as T
  }

  const response = await fetchWithErrorHandling(url)
  const json = await response.json()

  if (options?.cached) {
    jsonCache.set(url, json)
  }

  return json
}
