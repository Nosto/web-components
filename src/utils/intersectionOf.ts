export function intersectionOf(...arrays: string[][]) {
  if (arrays.length === 0) {
    return []
  }
  return arrays?.reduce((intersection, currentArray) => {
    return currentArray.filter(item => intersection.includes(item))
  })
}
