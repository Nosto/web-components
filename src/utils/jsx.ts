import type { Campaign } from "@/components/Campaign/Campaign"
import type { Control } from "@/components/Control/Control"
import type { DynamicCard } from "@/components/DynamicCard/DynamicCard"
import type { Image } from "@/components/Image/Image"
import type { Product } from "@/components/Product/Product"
import type { ProductCard } from "@/components/ProductCard/ProductCard"
import type { SectionCampaign } from "@/components/SectionCampaign/SectionCampaign"
import type { SimpleCard } from "@/components/SimpleCard/SimpleCard"
import type { SkuOptions } from "@/components/SkuOptions/SkuOptions"

type MaybeArray<T> = T | T[]

// remaps entries in GlobalEventHandlersEventMap to their respective React style event handlers
type GlobalEventHandlersMapping = {
  [K in keyof GlobalEventHandlersEventMap as `on${Capitalize<K>}`]?: (event: GlobalEventHandlersEventMap[K]) => void
}

type ElementMapping<T extends HTMLElement> = Partial<T> & GlobalEventHandlersMapping

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    type Element = HTMLElement
    type IntrinsicElements = {
      "nosto-campaign": ElementMapping<Campaign>
      "nosto-control": ElementMapping<Control>
      "nosto-dynamic-card": ElementMapping<DynamicCard>
      "nosto-image": ElementMapping<Image>
      "nosto-product": ElementMapping<Product>
      "nosto-product-card": ElementMapping<ProductCard>
      "nosto-section-campaign": ElementMapping<SectionCampaign>
      "nosto-simple-card": ElementMapping<SimpleCard>
      "nosto-sku-options": ElementMapping<SkuOptions>
      // Keep generic fallback for other HTML elements
      [key: string]: Record<string, unknown> & GlobalEventHandlersMapping
    }
  }
}

type Props = Record<string, unknown>
type Type = string | ((props: Props) => HTMLElement)
type Child = unknown
type Children = Array<Child>

const aliases: Record<string, string> = {
  className: "class",
  htmlFor: "for"
}

/**
 * Create an HTML element based on the given JSX type, props and children
 */
export function createElement(type: Type, props: Props, ...children: Children): HTMLElement {
  if (typeof type === "function") {
    return children?.length ? type({ ...props, children }) : type(props)
  }

  const element = document.createElement(type)
  applyProperties(element, props ?? {})
  children?.forEach(child => appendChild(element, child))
  return element
}

/**
 * Create a HTML fragment based on the given JSX children
 */
export function createFragment(arg: { children: Children } | null): DocumentFragment {
  const { children, ...props } = arg ?? { children: [] }
  const fragment = document.createDocumentFragment()
  Object.assign(fragment, props)
  children?.forEach(child => appendChild(fragment, child))
  return fragment
}

function isEventListener(key: string, value: unknown): value is EventListener {
  return key.startsWith("on") && typeof value === "function"
}

function toKebabCase(str: string) {
  return str.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase()
}

function applyProperties(element: HTMLElement, props: Props) {
  Object.entries(props).forEach(([key, value]) => {
    if (isEventListener(key, value)) {
      element.addEventListener(key.slice(2).toLowerCase(), value)
    } else if (key === "style") {
      Object.assign(element.style, value)
    } else {
      const normKey = aliases[key] ?? toKebabCase(key)
      element.setAttribute(normKey, String(value))
    }
  })
}

function appendChild(element: ParentNode, child: MaybeArray<Child>) {
  if (Array.isArray(child)) {
    child.forEach(c => appendChild(element, c))
  } else if (child instanceof DocumentFragment) {
    Array.from(child.children).forEach(c => element.appendChild(c))
  } else if (element instanceof HTMLTemplateElement) {
    appendChild(element.content, child)
  } else if (child instanceof HTMLElement) {
    element.appendChild(child)
  } else if (child !== undefined && child !== null && child !== false) {
    element.appendChild(document.createTextNode(String(child)))
  }
}