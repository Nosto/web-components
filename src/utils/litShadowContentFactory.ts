import { render, type TemplateResult } from "lit"

export function litShadowContentFactory(styles: string) {
  let cachedStyleSheet: CSSStyleSheet | null = null

  return async (element: HTMLElement, template: TemplateResult) => {
    if ("adoptedStyleSheets" in element.shadowRoot!) {
      if (!cachedStyleSheet) {
        cachedStyleSheet = new CSSStyleSheet()
        await cachedStyleSheet.replace(styles)
      }
      element.shadowRoot!.adoptedStyleSheets = [cachedStyleSheet]
      render(template, element.shadowRoot!)
    } else {
      // For browsers that don't support adoptedStyleSheets, we need to render the template
      // to a temporary element to get the HTML string, then inject it with styles
      const tempDiv = document.createElement("div")
      render(template, tempDiv)
      element.shadowRoot!.innerHTML = `
        <style>${styles}</style>
        ${tempDiv.innerHTML}
      `
    }
  }
}
