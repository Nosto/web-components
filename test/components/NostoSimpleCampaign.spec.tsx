/** @jsx createElement */
import { describe, it, expect, beforeAll, beforeEach, afterEach } from "vitest"
import { NostoSimpleCampaign } from "@/components/NostoSimpleCampaign/NostoSimpleCampaign"
import { mockNostoRecs } from "../mockNostoRecs"
import { addHandlers } from "../msw.setup"
import { createElement } from "../utils/jsx"
import { http, HttpResponse } from "msw"

describe("NostoSimpleCampaign", () => {
  let campaign: NostoSimpleCampaign

  beforeAll(() => {
    if (!customElements.get("nosto-simple-campaign")) {
      customElements.define("nosto-simple-campaign", NostoSimpleCampaign)
    }
  })

  beforeEach(() => {
    document.body.innerHTML = ""
    // Add MSW handlers for dynamic card requests
    addHandlers(
      http.get("/products/:handle", ({ params }) => {
        const handle = params.handle as string
        return HttpResponse.text(`<div class="product-card">${handle}</div>`, { status: 200 })
      })
    )
  })

  afterEach(() => {
    campaign?.remove?.()
  })

  it("should be defined as a custom element", () => {
    expect(customElements.get("nosto-simple-campaign")).toBeDefined()
  })

  it("should throw in connectedCallback if placement is missing", async () => {
    campaign = (<nosto-simple-campaign />) as NostoSimpleCampaign
    await expect(campaign.connectedCallback()).rejects.toThrow(
      "placement attribute is required for NostoSimpleCampaign"
    )
  })

  it("should load campaign and render in grid mode by default", async () => {
    const mockResult = {
      products: [
        { name: "Product 1", price: "$10", image_url: "https://example.com/img1.jpg", url: "/products/product-1" },
        { name: "Product 2", price: "$20", image_url: "https://example.com/img2.jpg", url: "/products/product-2" }
      ]
    }
    const { mockBuilder } = mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    expect(mockBuilder.load).toHaveBeenCalledWith()
    expect(campaign.querySelector(".nosto-grid")).toBeTruthy()
    expect(campaign.querySelectorAll(".nosto-product")).toHaveLength(2)
    expect(campaign.hasAttribute("loading")).toBe(false)
  })

  it("should render campaign in carousel mode when mode=carousel", async () => {
    const mockResult = {
      products: [{ name: "Product 1", price: "$10", url: "/products/product-1" }]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" mode="carousel" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    expect(campaign.querySelector(".nosto-carousel")).toBeTruthy()
    expect(campaign.querySelector(".nosto-grid")).toBeFalsy()
  })

  it("should render campaign in bundle mode when mode=bundle", async () => {
    const mockResult = {
      products: [{ name: "Product 1", price: "$10", url: "/products/product-1" }]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" mode="bundle" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    expect(campaign.querySelector(".nosto-bundle")).toBeTruthy()
    expect(campaign.querySelector(".nosto-grid")).toBeFalsy()
  })

  it("should handle empty campaign results gracefully", async () => {
    const { mockBuilder } = mockNostoRecs({ "test-placement": { products: [] } })

    campaign = (<nosto-simple-campaign placement="test-placement" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    expect(mockBuilder.load).toHaveBeenCalledWith()
    expect(campaign.innerHTML).toBe("")
    expect(campaign.hasAttribute("loading")).toBe(false)
  })

  it("should use NostoDynamicCard when card attribute is provided", async () => {
    const mockResult = {
      products: [{ name: "Product 1", price: "$10", handle: "test-handle" }]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" card="product-card" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    const dynamicCard = campaign.querySelector("nosto-dynamic-card")
    expect(dynamicCard).toBeTruthy()
    expect(dynamicCard?.getAttribute("handle")).toBe("test-handle")
    expect(dynamicCard?.getAttribute("template")).toBe("product-card")
  })

  it("should fall back to basic product display when card attribute is provided but no handle found", async () => {
    const mockResult = {
      products: [{ name: "Product 1", price: "$10" }]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" card="product-card" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    expect(campaign.querySelector("nosto-dynamic-card")).toBeFalsy()
    expect(campaign.querySelector(".nosto-product")).toBeTruthy()
    expect(campaign.querySelector(".nosto-product")?.textContent).toContain("Product 1")
  })

  it("should handle loading state correctly", async () => {
    const mockResult = { products: [{ name: "Product 1" }] }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    const loadingPromise = campaign.connectedCallback()
    expect(campaign.hasAttribute("loading")).toBe(true)

    await loadingPromise
    expect(campaign.hasAttribute("loading")).toBe(false)
  })

  it("should create NostoDynamicCard elements when card and handle are provided", async () => {
    const mockResult = {
      products: [
        { name: "Product 1", handle: "test-handle-1" },
        { name: "Product 2", handle: "test-handle-2" }
      ]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" card="product-card" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    const dynamicCards = campaign.querySelectorAll("nosto-dynamic-card")
    expect(dynamicCards).toHaveLength(2)
    expect(dynamicCards[0]?.getAttribute("handle")).toBe("test-handle-1")
    expect(dynamicCards[1]?.getAttribute("handle")).toBe("test-handle-2")
    expect(dynamicCards[0]?.getAttribute("template")).toBe("product-card")
  })

  it("should fallback to basic product elements when card is provided but no handle", async () => {
    const mockResult = {
      products: [
        { name: "Product 1", price: "$10" }, // no handle
        { name: "Product 2", handle: "test-handle" } // has handle
      ]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" card="product-card" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    const dynamicCards = campaign.querySelectorAll("nosto-dynamic-card")
    const basicProducts = campaign.querySelectorAll(".nosto-product")

    expect(dynamicCards).toHaveLength(1) // Only product with handle
    expect(basicProducts).toHaveLength(1) // Only product without handle
    expect(dynamicCards[0]?.getAttribute("handle")).toBe("test-handle")
  })

  it("should handle products with complete information in basic fallback", async () => {
    const mockResult = {
      products: [
        {
          name: "Complete Product",
          price: "$25.99",
          image_url: "https://example.com/image.jpg"
        }
      ]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    const productElement = campaign.querySelector(".nosto-product")
    expect(productElement).toBeTruthy()
    expect(productElement?.innerHTML).toContain("Complete Product")
    expect(productElement?.innerHTML).toContain("$25.99")
    expect(productElement?.innerHTML).toContain('src="https://example.com/image.jpg"')
    expect(productElement?.innerHTML).toContain('alt="Complete Product"')
  })

  it("should handle products with minimal information in basic fallback", async () => {
    const mockResult = {
      products: [{ name: "Basic Product" }] // Only name, no price or image
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    const productElement = campaign.querySelector(".nosto-product")
    expect(productElement).toBeTruthy()
    expect(productElement?.innerHTML).toContain("Basic Product")
    expect(productElement?.innerHTML).not.toContain("<img")
    expect(productElement?.innerHTML).not.toContain("price")
  })

  it("should handle products with no name gracefully", async () => {
    const mockResult = {
      products: [{ price: "$10" }] // No name
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (<nosto-simple-campaign placement="test-placement" />) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    const productElement = campaign.querySelector(".nosto-product")
    expect(productElement).toBeTruthy()
    expect(productElement?.innerHTML).toContain("Unnamed Product")
    expect(productElement?.innerHTML).toContain("$10")
  })

  it("should render different modes correctly with mixed product types", async () => {
    const mockResult = {
      products: [
        { name: "Product 1", handle: "handle-1" },
        { name: "Product 2" } // no handle
      ]
    }

    // Test grid mode
    mockNostoRecs({ "test-placement": mockResult })
    campaign = (
      <nosto-simple-campaign placement="test-placement" mode="grid" card="product-card" />
    ) as NostoSimpleCampaign
    document.body.appendChild(campaign)
    await campaign.connectedCallback()

    expect(campaign.querySelector(".nosto-grid")).toBeTruthy()
    expect(campaign.querySelectorAll("nosto-dynamic-card")).toHaveLength(1)
    expect(campaign.querySelectorAll(".nosto-product")).toHaveLength(1)
    campaign.remove()

    // Test carousel mode
    mockNostoRecs({ "test-placement": mockResult })
    campaign = (
      <nosto-simple-campaign placement="test-placement" mode="carousel" card="product-card" />
    ) as NostoSimpleCampaign
    document.body.appendChild(campaign)
    await campaign.connectedCallback()

    expect(campaign.querySelector(".nosto-carousel")).toBeTruthy()
    expect(campaign.querySelectorAll("nosto-dynamic-card")).toHaveLength(1)
    expect(campaign.querySelectorAll(".nosto-product")).toHaveLength(1)
    campaign.remove()

    // Test bundle mode
    mockNostoRecs({ "test-placement": mockResult })
    campaign = (
      <nosto-simple-campaign placement="test-placement" mode="bundle" card="product-card" />
    ) as NostoSimpleCampaign
    document.body.appendChild(campaign)
    await campaign.connectedCallback()

    expect(campaign.querySelector(".nosto-bundle")).toBeTruthy()
    expect(campaign.querySelectorAll("nosto-dynamic-card")).toHaveLength(1)
    expect(campaign.querySelectorAll(".nosto-product")).toHaveLength(1)
  })

  it("should default to grid mode for unknown modes", async () => {
    const mockResult = {
      products: [{ name: "Product 1", price: "$10" }]
    }
    mockNostoRecs({ "test-placement": mockResult })

    campaign = (
      <nosto-simple-campaign placement="test-placement" mode={"unknown-mode" as "grid" | "carousel" | "bundle"} />
    ) as NostoSimpleCampaign
    document.body.appendChild(campaign)

    await campaign.connectedCallback()

    expect(campaign.querySelector(".nosto-grid")).toBeTruthy()
    expect(campaign.querySelector(".nosto-carousel")).toBeFalsy()
    expect(campaign.querySelector(".nosto-bundle")).toBeFalsy()
  })
})
