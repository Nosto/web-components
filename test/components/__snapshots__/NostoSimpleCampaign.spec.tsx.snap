// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NostoSimpleCampaign > should render in carousel mode when specified 1`] = `
"<div class="nosto-simple-campaign nosto-carousel"><div class="nosto-carousel-container"><div class="nosto-product-card">
        <div class="nosto-product-image">
          <img src="https://example.com/carousel.jpg" alt="Carousel Product">
        </div>
        <div class="nosto-product-info">
          <h3 class="nosto-product-title">Carousel Product</h3>
          <div class="nosto-product-price">$39.99</div>
        </div>
      </div></div></div>"
`;

exports[`NostoSimpleCampaign > should render in grid mode by default 1`] = `
"<div class="nosto-simple-campaign nosto-grid"><div class="nosto-grid-container"><div class="nosto-product-card">
        <div class="nosto-product-image">
          <img src="https://example.com/image1.jpg" alt="Test Product 1">
        </div>
        <div class="nosto-product-info">
          <h3 class="nosto-product-title">Test Product 1</h3>
          <div class="nosto-product-price">$19.99</div>
        </div>
      </div><div class="nosto-product-card">
        <div class="nosto-product-image">
          <img src="https://example.com/image2.jpg" alt="Test Product 2">
        </div>
        <div class="nosto-product-info">
          <h3 class="nosto-product-title">Test Product 2</h3>
          <div class="nosto-product-price">$29.99</div>
        </div>
      </div></div></div>"
`;
